httpclient/httpclient.go

```diff
@@ -183,28 +191,30 @@ func (c *HTTPConnectDialer) DialContext(ctx context.Context, network, address st
 		c.cacheH2Mu.Unlock()
 	}
 
-	rawConn, err := c.Dialer.DialContext(ctx, network, c.ProxyUrl.Host)
-	if err != nil {
-		return nil, err
-	}
-
+	var err error
+	var rawConn net.Conn
 	negotiatedProtocol := ""
 	switch c.ProxyUrl.Scheme {
 	case "http":
+		rawConn, err = c.Dialer.DialContext(ctx, network, c.ProxyUrl.Host)
+		if err != nil {
+			return nil, err
+		}
 	case "https":
+		var tlsConn TLSConn
 		if c.DialTLS != nil {
-			rawConn, negotiatedProtocol, err = c.DialTLS(rawConn)
+			tlsConn, err = c.DialTLS(network, c.ProxyUrl.Host)
 			if err != nil {
 				return nil, err
 			}
-			break
-		}
-		tlsConf := tls.Config{
-			NextProtos: []string{"h2", "http/1.1"},
-			ServerName: c.ProxyUrl.Hostname(),
-		}
+		} else {
+			tlsConf := tls.Config{
+				NextProtos: []string{"h2", "http/1.1"},
+				ServerName: c.ProxyUrl.Hostname(),
+			}
 
-		tlsConn := tls.Client(rawConn, &tlsConf)
+			tlsConn = tls.Client(rawConn, &tlsConf)
+		}
 		err = tlsConn.Handshake()
```

Comment:
I think it would be helpful to have a comment explaining that this call to `Handshake` is possible redundant (e.g. when using `utls.Roller`) but always safe (because `Handshake` is idempotent).

Reply:
Good point, will add.